net: ipv4: Add support for conflict detection (RFC 5227)

rlubos opened the pull request:
This PR implements (optional) IPv4 conflict detection, as specified in [RFC 5227](https://datatracker.ietf.org/doc/html/rfc5227).

The implementation is based on the solution we had in IPv4 autoconf module, implementing [RFC 3927](https://datatracker.ietf.org/doc/html/rfc3927), which was just a more specialized case of a conflict detection described in RFC 5227.

IPv4 autoconf and DHCP has been integrated with the conflict detection events, so that new address can be generated/requested when a conflict is reported.

rlubos committed:
net: shell: Add ACD events to the event monitor

Include new ACD events in the shell event monitor.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: conn_mgr: Add support for ACD events

Connection manager needs to monitor ACD events as well to determine
whether a preferred IPv4 address is available.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
tests: net: all: Enable IPv4 ACD in the test suite

Make sure ACD is enabled in the build-all configuration.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: config: Add support for IPv4 ACD

In case IPv4 conflict detection is enabled, monitor network events to
determine whether IPv4 address is ready to use or not, so that the
library returns only after the network setup is fully complete.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: dhcpv4: Add support for conflict detection

In case a conflict was detected on a DHCP-assigned address, send a
Decline message to the server and start over.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: ipv4: autoconf: Integrate with the ACD module

The autoconf module can now reuse generic address conflict detection,
which was added for all address types.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: ipv4: Implement IPv4 address conflict detection

Add support for IPv4 conflict detection, as specified in RFC 5227.
The new feature is optional and disabled by default.

Address conflict detection was implemented as a part of the IPv4
autoconf feature can be generalized to be available for all address
types.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: l2: ethernet: arp: Simplify ACD case

In case of ACD Probe/Announcement, all we need is to generate ARP
packet, we don't really want any cache entries to be created or searched
for. There was a bug, that a cache entry was created for the
Announcement sent, resulting in skipped ARP packet generation and
malformed packet being sent by the ACD module.

Therefore, simplify all this, by simply returning early in case of
conflict detection packets.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

jukkar commented on subsys/net/lib/shell/events.c:
```diff
@@ -256,6 +256,18 @@ static char *get_l3_desc(struct event_msg *msg,
+		*desc2 = "fail";
+		info = net_addr_ntop(AF_INET, msg->data, extra_info,
+				     extra_info_len);
+		break;
```
Small nit. We should/could add information about the ACD enable/disable status to `subsys/net/lib/shell/ipv4.c` which prints information about IPv4 configuration.

rlubos replied:
Good point, added.

jukkar reviewed (commented):
LGTM, minor fix needed for CI to pass.

rlubos committed:
net: shell: ipv4: Add information about ACD

Print information in IPv4 shell whether address conflict detection is
enabled or not.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
Just a rebase, hopefully CI issues should be resolved now.
